Add BigQuery cookbook (#274)
mozilla/firefox-data-docs: src/cookbooks/bigquery.md

# Accessing and Working with BigQuery

This guide will give you a quick introduction to working with data stored
in [BigQuery](https://cloud.google.com/bigquery/)

As we transition to [GCP](https://cloud.google.com) we will use BigQuery as our primary data warehouse and
SQL Query engine. BigQuery will eventually replace our previous SQL Query
Engines, Presto and Athena, and our Parquet data lake.
## Table of Contents
<!-- toc -->

# Querying Tables

## Projects, Datasets and Tables in BigQuery
In GCP a [project](https://cloud.google.com/resource-manager/docs/creating-managing-projects) is a way to organize cloud resources. We use multiple
projects to maintain our BigQuery [datasets](https://cloud.google.com/bigquery/docs/datasets-intro). BigQuery datasets are a top-level container used to organize and
control access to tables and views.

### Caveats

- The date partition field (e.g. `submission_date_s3`, `submission_date`) is mostly used as a partitioning column,

> but it has changed from __YYYYMMDD__ form to the more standards-friendly __YYYY-MM-DD__ form.

Reviewer: ```suggestion
but it has changed from `YYYYMMDD` form to the more standards-friendly `YYYY-MM-DD` form.
```

- Unqualified queries can become very costly very easily. We've placed restrictions on large tables from accidentally querying "all data for all time",

> namely that you must make use of the date partition fields for large tables (like main_summary or clients_daily).

Reviewer: ```suggestion
namely that you must make use of the date partition fields for large tables (like `main_summary` or `clients_daily`).
```

> - Please read _Query Optimizations_ section that contains advice on how to reduce cost and improve query performance.

Reviewer 2: Link not working

> - Redash BigQuery Datasources will have a 10TB data scanned limit per query. Please let us know if you run into issues!

Reviewer: Generally `s/Redash/Re:dash/` to make the spellchecker happy

Reviewer: ```suggestion
- Re:dash BigQuery data sources will have a 10 TB data scanned limit per query. Please let us know in `#fx-metrics` on Slack if you run into issues!
```

### Projects with BigQuery datasets

|Project|Dataset|Purpose|
|---|---|---|

> |moz-fx-data-derived-datasets|    |Imported parquet data, [BigQuery ETL](https://github.com/mozilla/bigquery-etl), Ad hoc analysis|

Reviewer: ```suggestion
|`moz-fx-data-derived-datasets`|    |Imported parquet data, [BigQuery ETL](https://github.com/mozilla/bigquery-etl), ad-hoc analysis|
```

> |   |analysis|User generated tables for analysis|

Reviewer: nit for spellchecker: put all the dataset names in backticks

``` bash
bq ls moz-fx-data-derived-datasets:analysis
```
Query a table
 ``` bash
 bq query --nouse_legacy_sql 'select count(*) from `moz-fx-data-derived-datasets.telemetry.main_summary_v4` where submission_date_s3 = "2019-03-01"'
 ```

Additional examples and documentation can be found [here](https://cloud.google.com/bigquery/docs/bq-command-line-tool).

## From Databricks
_Work in progress_

> ## Interfaces

Reviewer 3: Logically I think it makes more sense if we should move this section above the "From <tool>" sections, and link to the detailed info for each of the tools

> |moz-fx-data-shar-nonprod-efed| |Data produced by stage structured ingestion infrastructure|

Reviewer: ```suggestion
|`moz-fx-data-shar-nonprod-efed`| |Data produced by stage structured ingestion infrastructure|
```

``` sql

SELECT col1, col2 FROM `project.dataset.table`

```
An example query from [Clients Last Seen Reference](../datasets/bigquery/clients_last_seen/reference.md)

``` sql
SELECT
    submission_date,
    os,
    COUNT(*) AS count
FROM
    telemetry.clients_last_seen
WHERE
    submission_date >= DATE_SUB(CURRENT_DATE, INTERVAL 1 WEEK)
    AND days_since_seen = 0
GROUP BY
    submission_date,
    os
HAVING
    count > 10 -- remove outliers
    AND lower(os) NOT LIKE '%windows%'
ORDER BY
    os,
    submission_date DESC
```

> # Access

Reviewer 2: Access should come before 'Writing queries' ?

> - Unqualified queries can become very costly very easily. We've placed restrictions on large tables from accidentally querying "all data for all time",

Reviewer 3: It doesn't look like we explicitly state in this doc that BigQuery charges by data scanned, and that the data is columnar. Maybe we can add this to the intro paragraph?

> 

Reviewer 3: Consider adding a caveat that there are no other partitioning fields in bq versions of parquet datasets (e.g. `sample_id` is no longer a partitioning field and will not necessarily reduce data scanned)

    - project name will be provided for you when your account is provisioned.

### `bq` Examples
List tables in a BigQuery dataset
``` bash
bq ls moz-fx-data-derived-datasets:analysis
```
Query a table
 ``` bash
 bq query --nouse_legacy_sql 'select count(*) from `moz-fx-data-derived-datasets.telemetry.main_summary_v4` where submission_date_s3 = "2019-03-01"'
 ```

Additional examples and documentation can be found [here](https://cloud.google.com/bigquery/docs/bq-command-line-tool).

## From Databricks
_Work in progress_

## Interfaces
BigQuery datasets and tables can be accessed by the following methods:

- [Redash](https://sql.telemetry.mozilla.org/)
- [GCP BigQuery Console](https://console.cloud.google.com/bigquery)
    - For advanced use cases including managing query outputs, table management. Requires GCP access to be granted by Data Operations.
- [GCP BigQuery API Access](https://cloud.google.com/bigquery/docs/reference/rest/)
    - For advanced use cases including automated workloads, ETL, [BigQuery Storage API](https://cloud.google.com/bigquery/docs/reference/storage/). Requires GCP access to be granted by Data Operations.

>     - Allows access to BigQuery via [bq command-line tool](https://cloud.google.com/bigquery/docs/bq-command-line-tool)

Reviewer: ```suggestion
    - Allows access to BigQuery via [`bq` command-line tool](https://cloud.google.com/bigquery/docs/bq-command-line-tool)
```

# Query Optimizations

To improve query performance and minimize the cost associated with using BigQuery please see the following query optimizations:

> - Avoid `SELECT *` and query only the columns you need

Reviewer 2: Suggestion: Avoid `SELECT *` by selecting only the columns you need

    - Using `SELECT *` is the most expensive way to query data. When you use `SELECT *` _BigQuery does a full scan of every column in the table._

>     - Applying a `LIMIT` clause to a `SELECT *` query does not affect of data read. You are billed for reading all bytes in the entire table.

Reviewer: ```suggestion
    - Applying a `LIMIT` clause to a `SELECT *` query does not affect the amount of data read. You are billed for reading all bytes in the entire table.
```

    - If you are experimenting with data or exploring data, use one of the [data preview options](https://cloud.google.com/bigquery/docs/best-practices-costs#preview-data) instead of `SELECT *`.
        - Preview support is coming soon to BigQuery datasources in [Redash](https://sql.telemetry.mozilla.org/)

> - Limit the amount of data scanned using a date partition filter

Reviewer 2: Nit: Limit the amount of data scanned *by* using a date partition filter

    - Tables that are larger than 1TB will require that you provide a date partition filter as part of the query.
    - You will receive an error if you attempt to query a table that requires a partition filter.

>         - `Cannot query over table 'moz-fx-data-derived-datasets.telemetry.main_summary_v4' without a filter over column(s) 'submission_date_s3' that can be used for partition elimination`

Reviewer 2: maybe add an example or link to example of how to add a date partition filter

- Reduce data before using a JOIN

>     - Trim the data as early in the query as possible, before the query performs a JOIN. If you reduce data early in the processing cycle, shuffling and other complex operations only execute on the data that you need.

Reviewer 2: I'm not sure of the exact audience here, but some examples of how to reduce(decrease/lessen) a query's output set prior to joining may be helpful here. Also 'reduce' could also be confused with map/reduce.

Author: I think this would vary based on what data is being queried. Do you have suggestions on a generic example?

Reviewer 2: Ah I wasn't entirely was sure about what you meant by reducing data, so was asking for examples. https://cloud.google.com/bigquery/docs/best-practices-performance-compute suggests using the largest table first in a join.

I'm assuming you meant using intermediate tables/views or sub queries with multiple where clauses as a way of decreasing sides of a join, prior to the join itself.

Author: > Ah I wasn't entirely was sure about what you meant by reducing data, so was asking for examples. https://cloud.google.com/bigquery/docs/best-practices-performance-compute suggests using the largest table first in a join.

The optimization you've mentioned is _yet another_ optimization that users can do :). 

> I'm assuming you meant using intermediate tables/views or sub queries with multiple where clauses as a way of decreasing sides of a join, prior to the join itself.

Yes this. I am just going to add that instead of providing an actual example.

- Do not treat WITH clauses as prepared statements
    - WITH clauses are used primarily for readability because they are not materialized. For example, placing all your queries in WITH clauses and then running UNION ALL is a misuse of the WITH clause. If a query appears in more than one WITH clause, it executes in each clause.
- Use approximate aggregation functions

>     - If the SQL aggregation function you're using has an equivalent approximation function, the approximation function will yield faster query performance. For example, instead of using `COUNT(DISTINCT)`, use `APPROX_COUNT_DISTINCT()`.

Reviewer 2: nice, did not know about these
